### PR TITLE
[WIP] Add discrete/continuous support+more samplers

### PR DESCRIPTION
**Pull request by rrkarim**
addr #187 #171
PR should include:
- [x] Support for more samplers (the design should be discussed)
- [x] Support for compound step sampling (maybe it is hard one to implement because of over-reliance on batching in tfp)
- [x] Support for discrete var sampling
- [ ] Support `auto_assign_method` based on a set of certain criterias (like in `pymc3`)
- [ ] More reach trace for compound step
- [ ] More tests (they suck for now)

### Current user design of compound step for user looks like this:
```python
# the easiest one with automatic method assignment
# for each free var
model = ... # we define the model
pm.sample(model(), sampler_type="compound")
# user defines the methods for selected vars in model
from pymc4.mcmc.samplers import NUTS, RandomWalkM
pm.sample(model(), 
    sampler_type="compound", 
    sampler_methods=[
        ("var", RandomWalkM), 
        ("mean", NUTS)
    ], 
    xla=True)
```
There is currently support for categorical distribution mcmc sampling with custom state function, implemented in `pymc4/distributions/state_functions.py`, but more distributions should be tested and supported.
- [x] Categorical
- [ ] **should be added as its tested and supported

### CODE REVIEW

**Comment by rrkarim** [comment]
There are some problem with sampling discrete distributions with `xla` enabled. I will look at the issue closer.

File: pymc4/distributions/distribution.py [hunk 1]
```diff
@@ -2,8 +2,10 @@
 import copy
 from typing import Optional, Union, Any
 
+from absl import logging
```
Comment (fonnesbeck) [inline pymc4/distributions/distribution.py]: Do we need/want this as a dependency?

Comment (rrkarim) [reply pymc4/distributions/distribution.py]: Isn't the `tf.logging` replaced with `absl-py` in `TF 2.0`?

File: pymc4/distributions/distribution.py [hunk 2]
```diff
@@ -195,6 +197,19 @@ def batch_shape(self):
     def event_shape(self):
         return self._distribution.event_shape
 
+    @property
+    def grad_support(self, state=None):
+        if state is None:
+            state = self.sample()
+        try:
+            logging.set_verbosity(logging.ERROR)
```
Comment (rrkarim) [inline pymc4/distributions/distribution.py]: @fonnesbeck It hides additional tf grad errors here.

**Comment by rrkarim** [comment]
@junpenglao can you please provide some feedback on the issues #233, I've implemented the sampling of the models with categorical distribution (with support of `new_state_fn`) but the implementation still relies on `CompounStep`. There is a brief notebook that will test sampling.

**Comment by rrkarim** [comment]
Sorry for some notebooks edited, tests do not pass without it. There are also some issues with notebook testing, radom example doesn't work on my local machine.

**Comment by rrkarim** [comment]
So basically, I had two ways of designing sampler methods. This PR implementation is similar to `pymc3`, which for me is more clear from user perspective. Another idea was to have `MCMCSampler` class with all the ops like `sample, run_chains etc.` and sampler subclasses that defined the parameters like `kernel`, `adaptive_kernel`, `trace_fn` and etc. Then `_assign_default_methods` method should be defined in `MCMCSampler` class (it is defined now in `CompoundStep`) so that user doesn't pass `sampler_type` argument, but the samplers are assigned automatically to each free var according to certain criterias. Basically, all the sampler methods are now the type of compound step. What I don't like about this design choice is how the parametrization for each sampler method is handled. S

**Comment by fonnesbeck** [comment]
I assume `Unititled1.ipynb` is not supposed to be there.

**Pull request closed without merging**